lib/fizzy/capi.cpp

```diff
@@ -0,0 +1,129 @@
+template <typename DstT, typename SrcT>
+DstT bit_cast(SrcT x) noexcept
+{
+    DstT z;
+    static_assert(sizeof(x) == sizeof(z));
+    __builtin_memcpy(&z, &x, sizeof(x));
+    return z;
+}
+}  // namespace
+
+extern "C" {
+bool fizzy_validate(const uint8_t* wasm_binary, size_t wasm_binary_size)
+{
+    try
+    {
+        fizzy::parse({wasm_binary, wasm_binary_size});
+        return true;
+    }
+    catch (...)
+    {
+        return false;
+    }
+}
+
+struct fizzy_module
+{
+    fizzy::Module module;
+};
+
+fizzy_module* fizzy_parse(const uint8_t* wasm_binary, size_t wasm_binary_size)
+{
+    try
+    {
+        auto cmodule = std::make_unique<fizzy_module>();
+        cmodule->module = fizzy::parse({wasm_binary, wasm_binary_size});
+        return cmodule.release();
+    }
+    catch (...)
+    {
+        return nullptr;
+    }
+}
+
+void fizzy_free_module(fizzy_module* module)
+{
+    delete module;
+}
+
+struct fizzy_instance
+{
+    fizzy::Instance* instance;
+};
+
+fizzy_instance* fizzy_instantiate(fizzy_module* module,
+    const fizzy_external_function* imported_functions, uint32_t imported_functions_size)
+{
+    try
+    {
+        std::vector<fizzy::ExternalFunction> functions(imported_functions_size);
+        for (size_t imported_func_idx = 0; imported_func_idx < imported_functions_size;
+             ++imported_func_idx)
+        {
+            auto func = [cfunc = imported_functions[imported_func_idx]](fizzy::Instance& instance,
+                            fizzy::span<const fizzy::Value> args,
+                            int depth) -> fizzy::ExecutionResult {
+                fizzy_instance cinstance{&instance};
+                const auto cres = cfunc.function(cfunc.context, &cinstance,
+                    reinterpret_cast<const fizzy_value*>(args.data()),
+                    static_cast<uint32_t>(args.size()), depth);
+
+                if (cres.trapped)
+                    return fizzy::Trap;
+                else if (!cres.has_value)
+                    return fizzy::Void;
+                else
+                    return bit_cast<fizzy::Value>(cres.value);
+            };
+
+            // TODO get type from input array
+            auto func_type = module->module.imported_function_types[imported_func_idx];
+
+            functions[imported_func_idx] =
+                fizzy::ExternalFunction{std::move(func), std::move(func_type)};
+        }
+
+        auto instance = fizzy::instantiate(std::move(module->module), std::move(functions));
+
+        auto cinstance = std::make_unique<fizzy_instance>();
+        cinstance->instance = instance.release();
+
+        fizzy_free_module(module);
+        return cinstance.release();
+    }
+    catch (...)
+    {
+        return nullptr;
+    }
+}
+
+void fizzy_free_instance(fizzy_instance* instance)
+{
+    delete instance->instance;
+    delete instance;
+}
+
+fizzy_execution_result fizzy_execute(fizzy_instance* instance, uint32_t func_idx,
```

Comment:
I don't think depth should be part of the public API, @chfast ?

Reply:
I think it must be. My plan is to convert it to a more generic opaque "ThreadExecutionContext". This is a sneak peak: https://github.com/wasmx/fizzy/pull/529/commits/5f08ee277a7959252b9dfd5763ce66dfc6b89b5d.

Of course this is up to discussion.

Reply:
Yeah, we discussed it a bit when introducing `depth` in C++ API. My argument for it is that without it host functions don't have an option to respect the call depth limit (calls to imported functions would reset depth to 0). For example, this test would fail (go into infinite recursion): https://github.com/wasmx/fizzy/blob/84abe560eaf30ccb1a1164e3290d8effdba8ad24/test/unittests/execute_call_test.cpp#L506